network-libp2p/src/connection_pool/behaviour.rs

```diff
@@ -821,6 +793,89 @@ impl Behaviour {
             }
         }
     }
+
+    fn on_listen_failure(&mut self, send_back_addr: &Multiaddr) {
+        // Get IP from multiaddress if it exists.
+        let ip_info = self.get_ip_info_from_multiaddr(send_back_addr);
+
+        // Decrement IP counters if needed
+        if let Some(ip_info) = ip_info {
+            self.decrement_ip_counters(&ip_info);
+        }
+        self.limits.peer_count = self.limits.peer_count.saturating_sub(1);
+    }
+
+    fn reached_peer_limit(&mut self) -> Result<(), ConnectionDenied> {
+        self.limits.peer_count = self.limits.peer_count.saturating_add(1);
+
+        // Check for the maximum peer count limit
+        if self.limits.peer_count > self.config.peer_count_max {
+            debug!(
+                connections = self.limits.peer_count,
+                "Max peer connections limit reached"
+            );
+            return Err(ConnectionDenied::new(Error::MaxPeerConnectionsReached));
+        }
+
+        Ok(())
+    }
+
+    fn reached_ip_limit(&mut self, ip_info: &IpInfo) -> Result<(), ConnectionDenied> {
+        // Increment peer IP counter
+        let ip_count = self.limits.ip_count.entry(ip_info.ip).or_insert(0);
+        *ip_count = ip_count.saturating_add(1);
+
+        // Increment peer subnet IP counter
+        if let Some(subnet_ip) = ip_info.subnet_ip {
+            let subnet_count = self.limits.ip_subnet_count.entry(subnet_ip).or_insert(0);
+            *subnet_count = subnet_count.saturating_add(1);
+        }
+
+        // Check if peer IP limit is reached
+        if *ip_count > self.config.peer_count_per_ip_max {
+            // IP address
+            debug!(ip=%ip_info.ip, limit=self.config.peer_count_per_ip_max, "Max peer connections per IP limit reached");
+            return Err(ConnectionDenied::new(Error::MaxPeerPerIPConnectionsReached));
+        }
+
+        // Check if peer subnet IP limit is reached
+        if let Some(subnet_ip) = ip_info.subnet_ip {
+            if *self
+                .limits
+                .ip_subnet_count
+                .get(&subnet_ip)
+                .expect("must have subnet_ip entry")
+                > self.config.peer_count_per_subnet_max
```

Comment:
The order does matter as I want to bump both counters before I check the limits. But I introduced a temp var to get rid of the second lookup 👍 

```rust
// Increment peer subnet IP counter
let subnet_ip_count = if let Some(subnet_ip) = ip_info.subnet_ip {
    Some(...)
} else {
    None
};
```

Reply:
Yes I realized that both counters must be increased. Your solution works perfectly.

There is an alternative if the order of increasing counters and the order of returning maybe need not be the same. Like so:
```rust 
let ip_count = ...;

if let Some(subnet_ip) = ip_info.subnet_ip {
    let subnet_ip_count = ...;
    if *subnet_ip_count > self.config.peer_count_per_subnet_max {
        ...
        return Err(...);
    }
}

if *ip_count > self.config.peer_count_per_ip_max {
    ...
    return Err(...);
}
```
Then you dont need to keep it around. But that means it will return with the subnet_ip error before returning with the ip_count error.

Reply:
That being said, I have no preference in this, so feel free to keep it as you have it already.